File: heron/stmgr/src/cpp/util/rotating-map.cpp
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

#include "util/rotating-map.h"
#include <iostream>
#include <list>
#include "proto/messages.h"
#include "basics/basics.h"
#include "errors/errors.h"
#include "threads/threads.h"
#include "network/network.h"

namespace heron {
namespace stmgr {

RotatingMap::RotatingMap(sp_int32 _nbuckets) {
  for (sp_int32 i = 0; i < _nbuckets; ++i) {
    auto head = new std::unordered_map<sp_int64, sp_int64>();
    buckets_.push_back(head);
  }
}

RotatingMap::~RotatingMap() {
  while (!buckets_.empty()) {
    auto m = buckets_.front();
    buckets_.pop_front();
    delete m;
  }
}

void RotatingMap::rotate() {
  auto m = buckets_.back();
  buckets_.pop_back();
  delete m;
  buckets_.push_front(new std::unordered_map<sp_int64, sp_int64>());
}

void RotatingMap::create(sp_int64 _key, sp_int64 _value) {
  std::unordered_map<sp_int64, sp_int64>* m = buckets_.front();
  (*m)[_key] = _value;
}

bool RotatingMap::anchor(sp_int64 _key, sp_int64 _value) {
  for (auto iter = buckets_.begin(); iter != buckets_.end(); ++iter) {
    std::unordered_map<sp_int64, sp_int64>* m = *iter;
    if (m->find(_key) != m->end()) {
      sp_int64 current_value = (*m)[_key];
      sp_int64 new_value = current_value ^ _value;
      (*m)[_key] = new_value;
      return new_value == 0;
    }
  }
  return false;
}

bool RotatingMap::remove(sp_int64 _key) {
  for (auto iter = buckets_.begin(); iter != buckets_.end(); ++iter) {
    size_t removed = (*iter)->erase(_key);
    if (removed > 0) return true;
  }
  return false;
}

}  // namespace stmgr
}  // namespace heron

File: heron/simulator/src/java/org/apache/heron/simulator/utils/RotatingMap.java
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.heron.simulator.utils;

import java.util.HashMap;
import java.util.LinkedList;
import java.util.Map;

/**
 * Rotating Map maintains a list of unordered maps.
 * Every time a rotate is called, it drops the
 * last map and instantiates a new map at the head
 * of the list. The create operation adds elements
 * to the front map of the list. The anchor and remove
 * operation do their operations starting from the
 * front of the list to the back.
 */

public class RotatingMap {
  private final LinkedList<Map<Long, Long>> buckets = new LinkedList<>();

  // Creates a rotating map with nBuckets maps
  public RotatingMap(int nBuckets) {
    for (int i = 0; i < nBuckets; i++) {
      Map<Long, Long> head = new HashMap<>();
      buckets.addLast(head);
    }
  }

  // Deletes the last map on the list and
  // instantiates a new map at the front of the list
  public void rotate() {
    Map<Long, Long> m = buckets.removeLast();
    buckets.addFirst(new HashMap<Long, Long>());
  }

  // Adds an item to the map at the front of the list
  public void create(long key, long value) {
    buckets.getFirst().put(key, value);
  }

  // Starting from the front of the list,
  // it checks if the key already exists and if so
  // xors another entry to that. Returns true if the
  // xor turns zero. False otherwise
  public boolean anchor(long key, long value) {
    for (Map<Long, Long> m : buckets) {
      if (m.containsKey(key)) {
        long currentValue = m.get(key);
        long newValue = currentValue ^ value;
        m.put(key, newValue);

        return newValue == 0;
      }
    }
    return false;
  }

  // Starting from the front of the list, checks
  // if the key was present in the map and if so
  // removes it. Returns true if the key was removed
  // from some map. False otherwise.
  public boolean remove(long key) {
    for (Map<Long, Long> m : buckets) {
      if (m.remove(key) != null) {
        return true;
      }
    }
    return false;
  }
}
